Userland/Libraries/LibGLSL/Parser.cpp

```diff
@@ -0,0 +1,1142 @@
+/*
+ * Copyright (c) 2021, Itamar S. <itamar8910@gmail.com>
+ * Copyright (c) 2023, Volodymyr V. <vvmposeydon@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Parser.h"
+#include "AST.h"
+#include <LibGLSL/Lexer.h>
+
+namespace GLSL {
+
+Parser::Parser(Vector<Token> tokens, String const& filename)
+    : m_filename(move(filename))
+    , m_tokens(move(tokens))
+{
+}
+
+ErrorOr<NonnullRefPtr<TranslationUnit>> Parser::parse()
+{
+    if (m_tokens.is_empty())
+        return create_root_ast_node({}, {});
+    auto unit = create_root_ast_node(m_tokens.first().start(), m_tokens.last().end());
+    unit->set_declarations(TRY(parse_declarations_in_translation_unit(*unit)));
+    return unit;
+}
+
+bool Parser::eof() const
+{
+    return m_state.token_index >= m_tokens.size();
+}
+
+void Parser::print_tokens() const
+{
+    for (auto& token : m_tokens) {
+        outln("{}", token.to_string());
+    }
+}
+
+ErrorOr<Optional<Parser::DeclarationType>> Parser::match_declaration_in_translation_unit()
+{
+    if (TRY(match_variable_declaration()))
+        return DeclarationType::Variable;
+    if (TRY(match_function_declaration()))
+        return DeclarationType::Function;
+    if (TRY(match_struct_declaration()))
+        return DeclarationType::Struct;
+    return Optional<DeclarationType>();
+}
+
+ErrorOr<bool> Parser::match_struct_declaration()
+{
+    save_state();
+    ScopeGuard state_guard = [this] { load_state(); };
```

Comment:
You seem to be using this in a few places, maybe make a full RAII helper for this

Reply:
I don't feel like it's really worth the trouble, plus there are at least two different versions of this (normal and armed scope guard) and more might come in the future.